File: src/ruby/tormenta20/models/material_especial.rb
# frozen_string_literal: true

require "active_record"
require "json"

module Tormenta20
  module Models
    # Model representing a MaterialEspecial (Special Material) in Tormenta20.
    #
    # Special materials can be used to craft weapons and armor,
    # providing additional effects and modifying the item's price.
    #
    # @example Find all special materials
    #   Tormenta20.materiais_especiais.all
    #
    # @example Find a specific material
    #   Tormenta20.materiais_especiais.find("adamante")
    #
    # @!attribute [rw] id
    #   @return [String] Unique identifier (e.g., "adamante")
    # @!attribute [rw] name
    #   @return [String] Material name in Portuguese
    # @!attribute [rw] description
    #   @return [String, nil] Material description
    # @!attribute [rw] applicable_to
    #   @return [Array<String>, nil] Item types this material can be applied to
    # @!attribute [rw] price_modifier
    #   @return [String, nil] Price modifier when using this material
    # @!attribute [rw] effects
    #   @return [Array<String>, nil] Effects provided by this material
    class MaterialEspecial < Base
      self.table_name = "materiais_especiais"

      validates :id, presence: true, uniqueness: true
      validates :name, presence: true

      # Convert the special material to a hash representation.
      #
      # @return [Hash] Hash with all material attributes (nil values excluded)
      def to_h
        {
          id: id,
          name: name,
          description: description,
          applicable_to: applicable_to,
          price_modifier: price_modifier,
          effects: effects
        }.compact
      end
    end
  end
end

File: src/ts/models/material_especial.ts
import { BaseModel } from "./base.js"
import { Query } from "../query.js"

const TABLE = "materiais_especiais"

/**
 * A Material Especial (special material) that equipment can be crafted from.
 *
 * @example
 * ```ts
 * MaterialEspecial.all()
 * MaterialEspecial.find("adamante")
 * ```
 */
export class MaterialEspecial extends BaseModel {
  get description(): string | null { return (this._row.description as string) ?? null }
  /** JSON array of equipment types this material can be used for. */
  get applicableTo(): unknown[] { return BaseModel.json(this._row.applicable_to, []) }
  /** JSON object with price modifier data (multiplier or flat bonus). */
  get priceModifier(): Record<string, unknown> { return BaseModel.json(this._row.price_modifier, {}) }
  /** JSON object with mechanical effects granted by the material. */
  get effects(): Record<string, unknown> { return BaseModel.json(this._row.effects, {}) }

  toH() {
    return {
      id: this.id, name: this.name, description: this.description,
      applicable_to: this.applicableTo, price_modifier: this.priceModifier, effects: this.effects,
    }
  }

  static query(): Query<MaterialEspecial> { return BaseModel.makeQuery(TABLE, MaterialEspecial) }
  static all(): MaterialEspecial[] { return MaterialEspecial.query().all() }
  static find(id: string): MaterialEspecial | null { return MaterialEspecial.query().find(id) }
  static first(): MaterialEspecial | null { return MaterialEspecial.query().first() }
  static last(): MaterialEspecial | null { return MaterialEspecial.query().last() }
  static count(): number { return MaterialEspecial.query().count() }
}
